cpu/rpx0xx/pio/i2c/i2c.c

```diff
@@ -0,0 +1,415 @@
+/*
+ * Copyright (C) 2021 Otto-von-Guericke Universität Magdeburg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     cpu_rpx0xx
+ * @{
+ *
+ * @file
+ * @brief       PIO program to emulate an I2C interface
+ *
+ * @author      Fabian Hüßler <fabian.huessler@ovgu.de>
+ *
+ * @}
+ */
+#include <errno.h>
+#include "byteorder.h"
+#include "periph_conf.h"
+#include "periph/gpio.h"
+#include "periph/pio/i2c.h"
+#include "pio/pio.h"
+#include "i2c_set_scl_sda.pio.h"
+#include "i2c.pio.h"
+
+#define I2C_PIO_IRQN        0 /* use irq flag 0 */
+#define I2C_PIO_IRQN_SM(sm) ((1u << I2C_PIO_IRQN) << (sm))
+
+#define INSTR_I2C_SCL0_SDA0 (((pio_instr_t[])PIO_SET_SCL_SDA_PROGRAM)[0])
+#define INSTR_I2C_SCL0_SDA1 (((pio_instr_t[])PIO_SET_SCL_SDA_PROGRAM)[1])
+#define INSTR_I2C_SCL1_SDA0 (((pio_instr_t[])PIO_SET_SCL_SDA_PROGRAM)[2])
+#define INSTR_I2C_SCL1_SDA1 (((pio_instr_t[])PIO_SET_SCL_SDA_PROGRAM)[3])
+
+#define I2C_INSTR_FRAME(icount) ((uint32_t)(((uint32_t)(icount)) << 10))
+#define I2C_INSTR(instr) ((uint32_t)(instr))
+#define I2C_DATA_FRAME(final, data, nak)        \
+    ((uint32_t)((((uint32_t)(final)) << 9) |    \
+                (((uint32_t)(data)) << 1)  |    \
+                 ((uint32_t)(nak))))
+
+enum {
+    I2C_REPSTART = 0x40 /* make sure this flag is not used by the I2C driver */
+};
+
+static volatile uint32_t _irq[PIO_NUMOF];
```

Comment:
```suggestion
static volatile uint32_t _irq[PIO_I2C_NUMOF];
```
?

Reply:
This comment made me realize that I did a mistake. `PIO_I2C_NUMOF` is the number of configured PIO I2C buses.
One bus is managed by one PIO state machine because it executes the PIO I2C program.
But a program can be executed by multiple state machines. So the program is shared between state machines.

So `PIO_I2C_NUMOF` can be grater than `PIO_NUMOF`..
The interrupts and program location exist per PIO and the state machine which issued the interrupt is encoded in the interrupt   mask. PIO interrupt and their handling is not well explained in the datasheet in my opinion.